### ISSUE TITLE
Mesh registration

### ISSUE THREAD

**Issue by Theophile88** [jeanfeydy/geomloss#15]
Bonjour Jean,

First of all, thanks a lot for sharing this wonderful work with the community.  Great job!

I am trying to use OT for mesh registration. I basically use the code of your tutorial here: https://www.kernel-operations.io/geomloss/_auto_examples/optimal_transport/plot_interpolation_3D.html

I just start from my own template mesh and keep track of the mesh triangulation. Note that the target and template are not closed meshes.

Here is the template (black) and target meshes before registration:
![template_targ_comp](https://user-images.githubusercontent.com/7260483/70144893-2be10600-169f-11ea-80d2-9504cca2c066.gif)

The final match is given here:

![match_comp](https://user-images.githubusercontent.com/7260483/70145150-ba558780-169f-11ea-9e95-8596c2d0572c.gif)

And the taget here:
![target_comp](https://user-images.githubusercontent.com/7260483/70146192-fe498c00-16a1-11ea-8871-a544bb60400c.gif)

As you can see the result are not good between the thumb and the 4 fingers. Would you be able to tell me if I (probably) did something wrong or if this kind of failure could be expected with OT? In the first case, any idea of what is wrong and how to solve it? (I know it quite an open question! :))

I tried to add a mesh constraint in the objective function (just add a L2 norm which penalizes changes in the edge lengths), but it does not solve this "split" problem.

I also tried to use landmarks by adding a template/target L2 mismatch term in the OF, but it did not work well. Is there a better way to use landmarks with the sinkhorn loss?

Thanks a lot in advance and do not hesitate to answer "Sorry dude but it could be anything and I do not have time and enough info to figure out what is wrong" :)

Merci,
Théophile

**Comment by jeanfeydy** [comment]
Hi @Theophile88 ,

Thanks for you kind remarks :-)

This type of problem is absolutely expected with OT theory "out-of-the-box" on simple spatial features (x,y,z): **OT = generalized sorting**, no more, no less, and cannot guarantee the preservation of topology. 
Fortunately though, as you'd expect, there are (good) workarounds.

**The first one is to work in a higher-quality feature space:** instead of representing your shapes as measures (= weighted point clouds) in R^3 (features = (x,y,z)), you can often work with unit normals and curvatures efficiently. In your case, I'd be very confident of getting good results with (x,y,z, mean curvature) features. For a full "measure-theoretic" analysis of these tools, see e.g.:
- [this paper](http://openaccess.thecvf.com/content_cvpr_2017/papers/Kaltenmark_A_General_Framework_CVPR_2017_paper.pdf) and [this thesis](https://tel.archives-ouvertes.fr/tel-00942078/document) for position+**orientation** features (*varifolds*),
- [this paper](https://link.springer.com/content/pdf/10.1007%2Fs10851-019-00888-x.pdf) and [this thesis](https://tel.archives-ouvertes.fr/tel-02180601/document) for position+orientation+**curvature** features (*normal cycles*).

(Note that these papers were written with a focus on kernel metrics, but can be readily generalized to OT losses.)

**The second is to combine OT with a topology-aware deformation model.** The simplest way of doing so is to use a linear Thin Plate Spline model, as in [this classic paper](https://www.cise.ufl.edu/~anand/pdf/rangarajan_cviu.pdf) - see e.g. the excellent albeit slightly outdated [Wikipedia page on point set registration](https://en.wikipedia.org/wiki/Point_set_registration#Thin_plate_spline_robust_point_matching). Going further, you can use more advanced diffeomorphic models, as in [my first Miccai paper](https://arxiv.org/pdf/1706.05218.pdf) (most likely overkill for your use-case). Note that these works were published before we fully understood how to de-bias, symmetrize and implement efficiently the Sinkhorn divergence: with the GeomLoss implementation, things are now much faster and robust.


Making all these tools easily available to the community through simple PyTorch layers is my main priority. 
Unfortunately though, even if implementing this stuff is fairly straightforward once you know the maths, documenting+packaging+testing everything takes a huge amount of time. With my close collaborators @joanglaunes, @bcharlier and the [Aramis Inria team](https://github.com/aramis-lab) who maintain the [Deformetrica software](http://www.deformetrica.org/), we prioritized a bottom-up development strategy:
1. The [KeOps library](http://www.kernel-operations.io/) implements all the low-level routines we need (and much more!). After two years of work, it is now a mature project.
2. My **PhD thesis** (available in January... I've been working on it full-time since July, and the open release is *finally* getting in sight) provides a pragmatic introduction to the field: GPU programming, kernel/OT loss functions, feature spaces, deformation models. As you'll see if you click on the links above, most of the literature on the topic is too math-heavy to be widely accessible: hopefully, a detailed tutorial on the topic will help readers to get quickly to the algorithms without having to worry about abstract geometric measure theory.
3. **GeomLoss** will provide robust geometric loss functions for images, volumes and meshes that take into account curvature and other features... But as of December 2019, due to my PhD thesis, I only had time to implement the very crude "SamplesLoss" layer for point clouds! Over the next few months (January-March 2020), I will definitely implement and document curvature-aware Loss functions that can handle your problem efficiently.
4. Finally, we will provide a collection of reference implementations for Splines, LDDMM and more advanced **deformation layers** in a separate repository. Long-term, the idea is to replace monolithic shape analysis softwares such as Deformetrica by a versatile collection of composable layers, glued together with PyTorch. My [personal shapes toolbox](https://plmlab.math.cnrs.fr/jeanfeydy/shapes_toolbox), used to generate quickly figures for papers, kind of gives you an intermediate midpoint between old-fashioned "big softwares" and the "agile packages" that we're targeting.

Coming back to this thread next summer, I should thus be able to give you a straightforward 2-line answer... But until then, I'm afraid that you will have to implement some of these things by yourself :-)

Hope that this helps,

Jean

**Comment by Theophile88** [comment]
Thanks, @jeanfeydy , for this quick and clear answer! This helps a lot.

I will try your suggestions and keep you posted.

Good luck with your PhD defense (although I do not think you need any with the quality of your work!) and let us know when your thesis is available! Looking forward to reading it!

**Comment by aluo-x** [comment]
I wonder if another possible alternative would be to regularize the mesh itself to ensure that the transformations (weakly) preserve some metric of the mesh itself.
